File: gfg/linked-list/set_2.cpp
#include <cstdlib>
#include <cstdio>

typedef struct node {
  int data;
  struct node *next;
} node;

// insert node at the beginning of the linked list
void push(node **headRef, int newData) {
  node *newNode = (node *) malloc(sizeof(node));
  newNode->data = newData;
  newNode->next = *headRef;
  *headRef = newNode;
}

// insert node after a given node
void insertAfter(node *prevNode, int newData) {
  if (prevNode == NULL) {
    printf("Error: prevNode cannot be NULL!");
    return;
  }
  node *newNode = (node *) malloc(sizeof(node));
  newNode->data = newData;
  newNode->next = prevNode->next;
  prevNode->next = newNode;
}

// insert node at the end of the linked list
void append(node **headRef, int newData) {
  node *newNode = (node *) malloc(sizeof(node));
  newNode->data = newData;
  newNode->next = NULL;
  node *last = *headRef;
  if (last == NULL) {
    *headRef = newNode;
    return;
  }
  while (last->next != NULL) {
    last = last->next;
  }
  last->next = newNode;
}

void printList(node *head) {
  while (head != NULL) {
    printf("%d ", head->data);
    head = head->next;
  }
}

int main() {
  node *head = NULL;
  // Insert 6.  So linked list becomes 6->NULL
  append(&head, 6);

  // Insert 7 at the beginning. So linked list becomes 7->6->NULL
  push(&head, 7);

  // Insert 1 at the beginning. So linked list becomes 1->7->6->NULL
  push(&head, 1);

  // Insert 4 at the end. So linked list becomes 1->7->6->4->NULL
  append(&head, 4);

  // Insert 8, after 7. So linked list becomes 1->7->8->6->4->NULL
  insertAfter(head->next, 8);

  printf("\n Created Linked list is: ");
  printList(head);

  return 0;
}

File: gfg/linked-list/set_2.java
class Node {
  int data;
  Node next;

  public Node(int data) {
    this.data = data;
    this.next = null;
  }
}

class LinkedList {
  Node head;

  public LinkedList() {
    this.head = null;
  }

  public void push(int newData) {
    Node newNode = new Node(newData);
    newNode.next = this.head;
    this.head = newNode;
  }

  public void insertAfter(Node prevNode, int newData) {
    if (prevNode == null) {
      System.out.println("Error: prevNode cannot be null");
      return;
    }
    Node newNode = new Node(newData);
    newNode.next = prevNode.next;
    prevNode.next = newNode;
  }

  public void append(int newData) {
    Node newNode = new Node(newData);
    Node last = this.head;
    if (last == null) {
      this.head = newNode;
      return;
    }
    while (last.next != null) {
      last = last.next;
    }
    last.next = newNode;
  }

  public void print() {
    Node tmp = this.head;
    while (tmp != null) {
      System.out.print(tmp.data + " ");
      tmp = tmp.next;
    }
    System.out.println();
  }
}

public class set_2 {
  public static void main(String[] args) {
    LinkedList list = new LinkedList();
    list.push(1);
    list.push(2);
    list.push(3);
    list.print();
    list.append(4);
    list.append(5);
    list.insertAfter(list.head.next, 6);
    list.print();
  }
}
